From 3a29d72ce4823b6f71ba4d43231ea1c315329b37 Mon Sep 17 00:00:00 2001
From: Howard Hsu <howard-yh.hsu@mediatek.com>
Date: Wed, 24 Sep 2025 09:48:29 +0800
Subject: [PATCH 251/254] mtk: hostapd: support link-specific add_btwt/del_btwt
 via link_id

Add support for link_id in add_btwt and del_btwt functions to enable
handling of multi-link devices, improving broadcast TWT service
management.

Signed-off-by: Howard Hsu <howard-yh.hsu@mediatek.com>
---
 src/ap/ap_drv_ops.h          | 18 ++++++++++++++----
 src/drivers/driver.h         |  4 ++--
 src/drivers/driver_nl80211.c | 12 ++++++++----
 3 files changed, 24 insertions(+), 10 deletions(-)

diff --git a/src/ap/ap_drv_ops.h b/src/ap/ap_drv_ops.h
index fbcf9532d..387913d9b 100644
--- a/src/ap/ap_drv_ops.h
+++ b/src/ap/ap_drv_ops.h
@@ -397,18 +397,28 @@ static inline int hostapd_drv_switch_color(struct hostapd_data *hapd,
 static inline int hostapd_drv_add_btwt(struct hostapd_data *hapd,
 				       struct broadcast_twt_param *param)
 {
+	s8 link_id = -1;
+
 	if (!hapd->driver || !hapd->driver->add_btwt || !hapd->drv_priv)
 		return -1;
-
-	return hapd->driver->add_btwt(hapd->drv_priv, param);
+#ifdef CONFIG_IEEE80211BE
+	if (hapd->conf->mld_ap)
+		link_id = hapd->mld_link_id;
+#endif
+	return hapd->driver->add_btwt(hapd->drv_priv, param, link_id);
 }
 
 static inline int hostapd_drv_del_btwt(struct hostapd_data *hapd, u8 id)
 {
+	s8 link_id = -1;
+
 	if (!hapd->driver || !hapd->driver->del_btwt || !hapd->drv_priv)
 		return -1;
-
-	return hapd->driver->del_btwt(hapd->drv_priv, id);
+#ifdef CONFIG_IEEE80211BE
+	if (hapd->conf->mld_ap)
+		link_id = hapd->mld_link_id;
+#endif
+	return hapd->driver->del_btwt(hapd->drv_priv, id, link_id);
 }
 #endif /* CONFIG_IEEE80211AX */
 
diff --git a/src/drivers/driver.h b/src/drivers/driver.h
index 66adc5c57..b3a13b8c7 100644
--- a/src/drivers/driver.h
+++ b/src/drivers/driver.h
@@ -4851,12 +4851,12 @@ struct wpa_driver_ops {
 	/**
 	* add_btwt - add new broadcast twt service period
 	*/
-	int (*add_btwt)(void *priv, struct broadcast_twt_param *param);
+	int (*add_btwt)(void *priv, struct broadcast_twt_param *param, s8 link_id);
 
 	/**
 	* del_btwt - delete broadcast twt service period
 	*/
-	int (*del_btwt)(void *priv, u8 id);
+	int (*del_btwt)(void *priv, u8 id, s8 link_id);
 
 	/**
 	 * add_tx_ts - Add traffic stream
diff --git a/src/drivers/driver_nl80211.c b/src/drivers/driver_nl80211.c
index 335c51faa..b762692f6 100644
--- a/src/drivers/driver_nl80211.c
+++ b/src/drivers/driver_nl80211.c
@@ -12315,7 +12315,7 @@ error:
 	return ret;
 }
 
-int nl80211_add_btwt(void *priv, struct broadcast_twt_param *param)
+int nl80211_add_btwt(void *priv, struct broadcast_twt_param *param, s8 link_id)
 {
 	struct i802_bss *bss = priv;
 	struct wpa_driver_nl80211_data *drv = bss->drv;
@@ -12342,7 +12342,9 @@ int nl80211_add_btwt(void *priv, struct broadcast_twt_param *param)
 	    (param->flow_type && nla_put_flag(msg, NL80211_BROADCAST_TWT_ATTR_FLOW_TYPE)) ||
 	    (param->recommendation == 4 &&
 	     nla_put_u8(msg, NL80211_BROADCAST_TWT_ATTR_DL_TID_BMP, param->rtwt.dl_tid_bmp) &&
-	     nla_put_u8(msg, NL80211_BROADCAST_TWT_ATTR_UL_TID_BMP, param->rtwt.ul_tid_bmp)))
+	     nla_put_u8(msg, NL80211_BROADCAST_TWT_ATTR_UL_TID_BMP, param->rtwt.ul_tid_bmp)) ||
+	    (link_id != NL80211_DRV_LINK_ID_NA &&
+	     nla_put_u8(msg, NL80211_BROADCAST_TWT_ATTR_LINK_ID, link_id)))
 		goto fail;
 
 	nla_nest_end(msg, data);
@@ -12359,7 +12361,7 @@ fail:
 	return ret;
 }
 
-int nl80211_del_btwt(void *priv, u8 id)
+int nl80211_del_btwt(void *priv, u8 id, s8 link_id)
 {
 	struct i802_bss *bss = priv;
 	struct wpa_driver_nl80211_data *drv = bss->drv;
@@ -12376,7 +12378,9 @@ int nl80211_del_btwt(void *priv, u8 id)
 	if (!data)
 		goto fail;
 
-	if (nla_put_u8(msg, NL80211_BROADCAST_TWT_ATTR_ID, id))
+	if (nla_put_u8(msg, NL80211_BROADCAST_TWT_ATTR_ID, id) ||
+	    (link_id != NL80211_DRV_LINK_ID_NA &&
+	     nla_put_u8(msg, NL80211_BROADCAST_TWT_ATTR_LINK_ID, link_id)))
 		goto fail;
 
 	nla_nest_end(msg, data);
-- 
2.45.2

